token_filter.module

```diff
@@ -64,9 +64,123 @@ function _token_filter_filter_tokens($text, $filter, $format, $langcode, $cache,
     $data['user'] = $user;
   }
 
-  return token_replace($text, $data, $options);
+  // The rest of this function comes from token_replace(), but we insert our own
+  // additional filtering of the tokens to be replaced.
+  $text_tokens = token_scan($text);
+  if (empty($text_tokens)) {
+    return $text;
+  }
+
+  $replace = $filter->settings['replace'];
+
+  // Convert the token list to a regular expression.
+  $to_replace = array(
+    '/(\r\n?|\n)/', // newlines
+    '/\\\\\*/', // asterisks
+  );
+  $replacements = array(
+    '|', // logical OR
+    '[a-zA-Z0-9_\-\:]*', // alphanumeric, hyphen, underscore, or colon
+  );
+  $tokens_quoted = preg_quote($filter->settings['tokens']);
+  $regexps = '/^' . preg_replace($to_replace, $replacements, $tokens_quoted) . '$/';
+
+  // Compare each found token against our list to find matches. If there's a
+  // match, see whether we replace the token and if not, remove it from the list.
+  foreach ($text_tokens as $type => $value) {
+    foreach ($value as $name => $token) {
+      $matched = (bool) preg_match($regexps, $token);
+      if (($matched && !$replace) || (!$matched && $replace)) {
+        unset($text_tokens[$type][$name]);
+        if (empty($text_tokens[$type])) {
+          unset($text_tokens[$type]);
+        }
+      }
+    }
+  }
+
+  // Continue with the default behavior and code of token_replace().
+  $replacements = array();
+  foreach ($text_tokens as $type => $tokens) {
+    $replacements += token_generate($type, $tokens, $data, $options);
+    if (!empty($options['clear'])) {
+      $replacements += array_fill_keys($tokens, '');
+    }
+  }
+
+  // Optionally alter the list of replacement values.
+  if (!empty($options['callback'])) {
+    $function = $options['callback'];
+    $function($replacements, $data, $options);
+  }
+
+  $tokens = array_keys($replacements);
+  $values = array_values($replacements);
+
+  return str_replace($tokens, $values, $text);
 }
 
+/**
+ * Implements callback_filter_settings().
+ *
+ * Filter settings callback for the token text filter.
+ */
+function _token_filter_settings($form, &$form_state, $filter, $format) {
+  $settings['replace'] = array(
+    '#type' => 'radios',
+    '#options' => array(
+      0 => t('Ignore these tokens'),
+      1 => t('Replace only these tokens'),
+    ),
+    '#default_value' => isset($filter->settings['replace']) ? $filter->settings['replace'] : 0,
+    '#required' => TRUE,
+  );
+  $settings['tokens'] = array(
+    '#type' => 'textarea',
+    '#title' => t('Tokens'),
+    '#default_value' => isset($filter->settings['tokens']) ? $filter->settings['tokens'] : '',
+    '#description' => t('Enter tokens to ignore or replace, one per line, including the brackets. You can use "*" as a wildcard, for example, [user:*]. Select "Ignore" above and leave this blank to replace all recognized tokens. Ignored tokens will be left unaltered in the text.'),
+  );
+  return $settings;
```

Comment:
Wondering if the token tree may be more helpful here, than in the function `_token_filter_filter_tips()`? You don't really need to see the tokens over there, but here, it'd be helpful to have a list. I realize the returned values of this function are shown in a modal window, but I have actually used a token tree inside a modal (in the module `token_block`) and it works. You get a second modal with the tree, which you can move to the side.

Also, wondering if it may be helpful to show the token tree for all core entity tokens here, since those are likely the ones you'll use. You would do this by passing the key 'token_types' => array('node', 'user', ...)  when you call `theme('token_tree'...)`. I see you are only passing `user`.

Reply:
I was only passing user because that was the only core entity that was guaranteed to be usable. There's not necessarily a node context available for `[node:?]` tokens to get rendered. Do you think it's better to list more tokens even if some of them may not be recognized, or to be conservative and leave out tokens that may not be recognized?

Reply:
To me, showing at least the `node` tokens is important. For example, `[node:author:mail]` could be a token I may want to forbid. 

BTW, I just found a bug generated by the new code I'll report in a second.